File: frontend/src/types/dataset.ts
export interface Dataset {
  id: string;
  name: string;
  filename: string;
  file_size: number | null;
  point_count: number | null;
  crs: string | null;
  status: "uploading" | "uploaded" | "processing" | "ready" | "failed";
  created_at: string;
  updated_at: string;
}

export interface Job {
  id: string;
  dataset_id: string;
  status: "pending" | "processing" | "complete" | "failed";
  progress: number;
  error: string | null;
  started_at: string | null;
  completed_at: string | null;
}

export interface ChunkUploadState {
  datasetId: string;
  filename: string;
  totalChunks: number;
  uploadedChunks: number;
  status: "uploading" | "assembling" | "done" | "error";
  error?: string;
}

File: api/app/models/dataset.py
import uuid
from datetime import datetime, timezone
from sqlalchemy import Column, String, BigInteger, Float, Text, DateTime, ForeignKey, Enum as SAEnum
from sqlalchemy.orm import relationship
from app.db import Base
import enum

class DatasetStatus(str, enum.Enum):
    uploading = "uploading"
    uploaded = "uploaded"
    processing = "processing"
    ready = "ready"
    failed = "failed"

class JobStatus(str, enum.Enum):
    pending = "pending"
    processing = "processing"
    complete = "complete"
    failed = "failed"

class Dataset(Base):
    __tablename__ = "datasets"
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    name = Column(String, nullable=False)
    filename = Column(String, nullable=False)
    file_size = Column(BigInteger, nullable=True)
    point_count = Column(BigInteger, nullable=True)
    crs = Column(String, nullable=True)
    bounds = Column(Text, nullable=True)
    status = Column(SAEnum(DatasetStatus), default=DatasetStatus.uploading, nullable=False)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))
    updated_at = Column(DateTime, default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))
    job = relationship("Job", back_populates="dataset", uselist=False)

class Job(Base):
    __tablename__ = "jobs"
    id = Column(String, primary_key=True, default=lambda: str(uuid.uuid4()))
    dataset_id = Column(String, ForeignKey("datasets.id"), nullable=False)
    status = Column(SAEnum(JobStatus), default=JobStatus.pending, nullable=False)
    progress = Column(Float, default=0.0)
    error = Column(Text, nullable=True)
    started_at = Column(DateTime, nullable=True)
    completed_at = Column(DateTime, nullable=True)
    dataset = relationship("Dataset", back_populates="job")
